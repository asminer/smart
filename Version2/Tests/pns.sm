
// $Id$

/* 
   Petri net test file.
*/

#MatrixByRows true

print("Simple first net to check:\n",
"              p2 --> t2 --> p4\n",
"             /                \\\n",
"            /                  \\\n",
"   p1 --> t1                    t5  ---> (p1)\n",
"	    \\                  /\n",
"             \\   --> t3 -->   /\n",
"              p3            p5\n",
"                 <-- t4 <--\n\n");

pn net1(int N) := {
  place p1, p2, p3, p4, p5, p6;

  trans t1, t2, t3, t4, t5;

  init(p1:N);

  arcs(
	p1 : t1,
	t1 : p2, t1 : p3,
        p2 : t2, t2 : p4,
        p3 : t3, t3 : p5,
	p5 : t4, t4 : p3,
	p4 : t5, p5 : t5, t5 : p1
  );

  int m1 := num_states(true);
  int m2 := num_arcs(true);
};

string states1(int n) := sprint(net1(n).m1, " states total");
string arcs1(int n) := sprint(net1(n).m2, " arcs total");

print("N=1, safe markings:\n");
#PnMarkingStyle SAFE

print("States:\n", states1(1), "\n", arcs1(1), "\n");

print("N=2, indexed markings:\n");
#PnMarkingStyle INDEXED

print("States:\n", states1(2), "\n", arcs1(2), "\n");

print("N=3, vector markings:\n");
#PnMarkingStyle VECTOR

print("States:\n", states1(3), "\n", arcs1(3), "\n");

print("Should be simultaneous RG/S construction:\n");
print(arcs1(1), "\n");

print("Clear: ", arcs1(0), "\n");
print("Lexical ordering test\n");
#StateDisplayOrder LEXICAL
print("States:\n", states1(1), "\n", arcs1(1), "\n");

print("Various PN warnings:\n");

pn net2 := {
  place p, p;
  trans t, t, t;
  print("Duplicate init warning should follow:\n");
  init(p:2, p:3);
  print("Duplicate arc warning should follow:\n");
  arcs(p:t, p:t);
  int m := num_states(false);
};
  
print(net2.m - net2.m, " should be 0\n");



