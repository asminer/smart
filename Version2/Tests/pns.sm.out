Simple first net to check:
              p2 --> t2 --> p4
             /                \
            /                  \
   p1 --> t1                    t5  ---> (p1)
	    \                  /
             \   --> t3 -->   /
              p3            p5
                 <-- t4 <--

N=1, safe markings:
States:
State 0: [p1]
State 1: [p2, p3]
State 2: [p3, p4]
State 3: [p2, p5]
State 4: [p4, p5]
5 states total
Reachability graph:
	From state 0
		To state 1
	From state 1
		To state 2
		To state 3
	From state 2
		To state 4
	From state 3
		To state 1
		To state 4
	From state 4
		To state 0
		To state 2
8 arcs total
N=2, indexed markings:
States:
State 0: [p1:2, p2:0, p3:0, p4:0, p5:0, p6:0]
State 1: [p1:1, p2:1, p3:1, p4:0, p5:0, p6:0]
State 2: [p1:0, p2:2, p3:2, p4:0, p5:0, p6:0]
State 3: [p1:1, p2:0, p3:1, p4:1, p5:0, p6:0]
State 4: [p1:1, p2:1, p3:0, p4:0, p5:1, p6:0]
State 5: [p1:0, p2:1, p3:2, p4:1, p5:0, p6:0]
State 6: [p1:0, p2:2, p3:1, p4:0, p5:1, p6:0]
State 7: [p1:1, p2:0, p3:0, p4:1, p5:1, p6:0]
State 8: [p1:0, p2:0, p3:2, p4:2, p5:0, p6:0]
State 9: [p1:0, p2:1, p3:1, p4:1, p5:1, p6:0]
State 10: [p1:0, p2:2, p3:0, p4:0, p5:2, p6:0]
State 11: [p1:0, p2:0, p3:1, p4:2, p5:1, p6:0]
State 12: [p1:0, p2:1, p3:0, p4:1, p5:2, p6:0]
State 13: [p1:0, p2:0, p3:0, p4:2, p5:2, p6:0]
14 states total
Reachability graph:
	From state 0
		To state 1
	From state 1
		To state 2
		To state 3
		To state 4
	From state 2
		To state 5
		To state 6
	From state 3
		To state 5
		To state 7
	From state 4
		To state 1
		To state 6
		To state 7
	From state 5
		To state 8
		To state 9
	From state 6
		To state 2
		To state 9
		To state 10
	From state 7
		To state 0
		To state 3
		To state 9
	From state 8
		To state 11
	From state 9
		To state 1
		To state 5
		To state 11
		To state 12
	From state 10
		To state 6
		To state 12
	From state 11
		To state 3
		To state 8
		To state 13
	From state 12
		To state 4
		To state 9
		To state 13
	From state 13
		To state 7
		To state 11
34 arcs total
N=3, vector markings:
States:
State 0: [3, 0, 0, 0, 0, 0]
State 1: [2, 1, 1, 0, 0, 0]
State 2: [1, 2, 2, 0, 0, 0]
State 3: [2, 0, 1, 1, 0, 0]
State 4: [2, 1, 0, 0, 1, 0]
State 5: [0, 3, 3, 0, 0, 0]
State 6: [1, 1, 2, 1, 0, 0]
State 7: [1, 2, 1, 0, 1, 0]
State 8: [2, 0, 0, 1, 1, 0]
State 9: [0, 2, 3, 1, 0, 0]
State 10: [0, 3, 2, 0, 1, 0]
State 11: [1, 0, 2, 2, 0, 0]
State 12: [1, 1, 1, 1, 1, 0]
State 13: [1, 2, 0, 0, 2, 0]
State 14: [0, 1, 3, 2, 0, 0]
State 15: [0, 2, 2, 1, 1, 0]
State 16: [0, 3, 1, 0, 2, 0]
State 17: [1, 0, 1, 2, 1, 0]
State 18: [1, 1, 0, 1, 2, 0]
State 19: [0, 0, 3, 3, 0, 0]
State 20: [0, 1, 2, 2, 1, 0]
State 21: [0, 2, 1, 1, 2, 0]
State 22: [0, 3, 0, 0, 3, 0]
State 23: [1, 0, 0, 2, 2, 0]
State 24: [0, 0, 2, 3, 1, 0]
State 25: [0, 1, 1, 2, 2, 0]
State 26: [0, 2, 0, 1, 3, 0]
State 27: [0, 0, 1, 3, 2, 0]
State 28: [0, 1, 0, 2, 3, 0]
State 29: [0, 0, 0, 3, 3, 0]
30 states total
Reachability graph:
	From state 0
		To state 1
	From state 1
		To state 2
		To state 3
		To state 4
	From state 2
		To state 5
		To state 6
		To state 7
	From state 3
		To state 6
		To state 8
	From state 4
		To state 1
		To state 7
		To state 8
	From state 5
		To state 9
		To state 10
	From state 6
		To state 9
		To state 11
		To state 12
	From state 7
		To state 2
		To state 10
		To state 12
		To state 13
	From state 8
		To state 0
		To state 3
		To state 12
	From state 9
		To state 14
		To state 15
	From state 10
		To state 5
		To state 15
		To state 16
	From state 11
		To state 14
		To state 17
	From state 12
		To state 1
		To state 6
		To state 15
		To state 17
		To state 18
	From state 13
		To state 7
		To state 16
		To state 18
	From state 14
		To state 19
		To state 20
	From state 15
		To state 2
		To state 9
		To state 20
		To state 21
	From state 16
		To state 10
		To state 21
		To state 22
	From state 17
		To state 3
		To state 11
		To state 20
		To state 23
	From state 18
		To state 4
		To state 12
		To state 21
		To state 23
	From state 19
		To state 24
	From state 20
		To state 6
		To state 14
		To state 24
		To state 25
	From state 21
		To state 7
		To state 15
		To state 25
		To state 26
	From state 22
		To state 16
		To state 26
	From state 23
		To state 8
		To state 17
		To state 25
	From state 24
		To state 11
		To state 19
		To state 27
	From state 25
		To state 12
		To state 20
		To state 27
		To state 28
	From state 26
		To state 13
		To state 21
		To state 28
	From state 27
		To state 17
		To state 24
		To state 29
	From state 28
		To state 18
		To state 25
		To state 29
	From state 29
		To state 23
		To state 27
88 arcs total
Should be simultaneous RG/S construction:
Reachability graph:
	From state 0
		To state 1
	From state 1
		To state 2
		To state 3
	From state 2
		To state 4
	From state 3
		To state 1
		To state 4
	From state 4
		To state 0
		To state 2
8 arcs total
Clear: WARNING:
	Assuming zero initial marking in PN net1
Reachability graph:
	From state 0
0 arcs total
Lexical ordering test
States:
State 0: [0, 0, 0, 1, 1, 0]
State 1: [0, 0, 1, 1, 0, 0]
State 2: [0, 1, 0, 0, 1, 0]
State 3: [0, 1, 1, 0, 0, 0]
State 4: [1, 0, 0, 0, 0, 0]
5 states total
Reachability graph:
	From state 0
		To state 1
		To state 4
	From state 1
		To state 0
	From state 2
		To state 0
		To state 3
	From state 3
		To state 1
		To state 2
	From state 4
		To state 3
8 arcs total
Various PN warnings:
ERROR in file pns.sm near line 70:
	Duplicate identifier p within model
ERROR in file pns.sm near line 71:
	Duplicate identifier t within model
ERROR in file pns.sm near line 71:
	Duplicate identifier t within model
Duplicate init warning should follow:
WARNING in file pns.sm near line 73:
	Summing duplicate initialization for place p in PN net2
Duplicate arc warning should follow:
WARNING in file pns.sm near line 75:
	Summing cardinalities on duplicate arc
	from p to t in PN net2
0 should be 0
